mdoc/Mono.Documentation/MDocUpdater.cs

```diff
@@ -3626,124 +3623,129 @@ static bool UpdateAssemblyVersionForAssemblyInfo (XmlElement e, XmlElement root,
+                                    if (formatter.TryGetAttributeString(customAttr, out string attrStr, prefix))
+                                    {
+                                        node = WriteElementText(ae, "AttributeName", attrStr, forceNewElement: true);
+                                        WriteElementAttribute(node, Language, formatter.Language);
+                                    }
+                                }
+                            }
+
+                            if (fx.FrameworksCount > 1)
+                            {
+                                ae.SetAttribute(Consts.FrameworkAlternate, fx.Name);
+                            }
+                            existingAttrElements.Add(csharpAttrStr, ae);
+                        }
+                    }
                 }
-            }
-
-            // iterate the state's attributes to find attributes that aren't 
-            // a part of this framework's attributes.
-            foreach(var stateAttribute in attributesState)
+            }
+
+            // iterate the state's attributes to find attributes that aren't 
+            // a part of this framework's attributes.
+            foreach (var stateAttribute in existingAttrElements)
             {
-                if (attributesCurrent.ContainsKey(stateAttribute.ExistingValue)) 
+                if (currentAttrBatch.Contains(stateAttribute.Key))
                 {
                     continue;
                 }
                 else // let's remove it 
                 {
-                    if (fx.FrameworksCount > 1) 
+                    if (fx.FrameworksCount > 1)
                     {
-                        var newfxlist = FXUtils.RemoveFXFromList (stateAttribute.XElement.GetAttribute (Consts.FrameworkAlternate), fx.Name);
-                        stateAttribute.XElement.SetAttribute (Consts.FrameworkAlternate, newfxlist);
+                        var newfxlist = FXUtils.RemoveFXFromList(stateAttribute.Value.GetAttribute(Consts.FrameworkAlternate), fx.Name);
+                        stateAttribute.Value.SetAttribute(Consts.FrameworkAlternate, newfxlist);
 
                     }
-                    else 
-                        stateAttribute.XElement.ParentNode.RemoveChild (stateAttribute.XElement);
+                    else
+                        stateAttribute.Value.ParentNode.RemoveChild(stateAttribute.Value);
                 }
             }
 
             // clean up
-            bool lastFxForAssembly = (!string.IsNullOrWhiteSpace (assemblyName) && fx.IsLastFrameworkForAssembly (assemblyName));
-            bool lastFxForType = (typeEntry != null && fx.IsLastFrameworkForType (typeEntry));
+            bool lastFxForAssembly = (isAssemblyAttribute && fx.IsLastFrameworkForAssembly(assemblyName));
+            bool lastFxForType = (typeEntry != null && fx.IsLastFrameworkForType(typeEntry));
             if (lastFxForAssembly || lastFxForType)
```

Comment:
In retrospect, this should definitely be clearer, because it might be prone to error ... this method should know whether we're adding assembly attributes, or type attributes (or really, any of the other attribute scenarios, like on parameters and return types)